Kernel/Devices/HID/KeyboardDevice.cpp

```diff
@@ -20,287 +20,514 @@
+        key_event.scancode = 0xe0f07ce0f012;
+    } else if (event.bytes_count == 8) {
+        // 0xE1, 0x14, 0x77, 0xE1, 0xF0, 0x14, 0xF0, 0x77 - pause pressed
+        auto first_byte_prefix = event.scan_code_bytes[0];
+        auto second_byte_prefix = event.scan_code_bytes[1];
+        auto third_byte_prefix = event.scan_code_bytes[2];
+        auto fourth_byte_prefix = event.scan_code_bytes[3];
+        auto fifth_byte_prefix = event.scan_code_bytes[4];
+        auto sixth_byte_prefix = event.scan_code_bytes[5];
+        auto seventh_byte_prefix = event.scan_code_bytes[6];
+        auto eight_byte_prefix = event.scan_code_bytes[7];
+        VERIFY(first_byte_prefix == 0xe1);
+        VERIFY(second_byte_prefix == 0x14);
+        VERIFY(third_byte_prefix == 0x77);
+        VERIFY(fourth_byte_prefix == 0xe1);
+        VERIFY(fifth_byte_prefix == 0xf0);
+        VERIFY(sixth_byte_prefix == 0x14);
+        VERIFY(seventh_byte_prefix == 0xf0);
+        VERIFY(eight_byte_prefix == 0x77);
+        key_event.code_entry = KeyCodeEntry { Key_PauseBreak, 0xFF };
+        key_event.scancode = 0xe11477e1f014f077;
     }
 
-    if ((m_modifiers & Mod_Alt) != 0 && key >= Key_1 && key <= Key_1 + ConsoleManagement::s_max_virtual_consoles + 1) {
-        // FIXME: Do something sanely here if we can't allocate a work queue?
-        MUST(g_io_work->try_queue([key]() {
-            ConsoleManagement::the().switch_to(key - Key_1);
-        }));
+    switch (key_event.code_entry.key_code) {
+    case Key_RightAlt:
+        update_modifier(Mod_AltGr, key_event.is_press());
+        break;
+    case Key_Alt:
+        update_modifier(Mod_Alt, key_event.is_press());
+        break;
+    case Key_Control:
+        update_modifier(Mod_Ctrl, key_event.is_press());
+        break;
+    case Key_Super:
+        m_left_super_pressed = key_event.is_press();
+        update_modifier(Mod_Super, m_left_super_pressed || m_right_super_pressed);
+        break;
+    case Key_LeftShift:
+        m_left_shift_pressed = key_event.is_press();
+        update_modifier(Mod_Shift, m_left_shift_pressed || m_right_shift_pressed);
+        break;
+    case Key_RightShift:
+        m_right_shift_pressed = key_event.is_press();
+        update_modifier(Mod_Shift, m_left_shift_pressed || m_right_shift_pressed);
+        break;
+    default:
+        break;
+    }
+
+    return key_event;
+}
+
+void KeyboardDevice::handle_scan_code_input_event(ScanCodeEvent event)
+{
+    RawKeyEvent raw_event {};
+    if (event.sent_scan_code_set == ScanCodeSet::Set1) {
+        raw_event = handle_scan_code_input_event_set1(event);
+    } else if (event.sent_scan_code_set == ScanCodeSet::Set2) {
+        raw_event = handle_scan_code_input_event_set2(event);
+    } else if (event.sent_scan_code_set == ScanCodeSet::Set3) {
+        // FIXME: Implement support for scan code set 3!
+        VERIFY_NOT_REACHED();
+    } else {
+        VERIFY_NOT_REACHED();
```

Comment:
How much is this Scan Code 1/2 translation logic necessary for the operation of a USB keyboard? If we want `KeyboardDevice` to be a generic device used by both PS/2 and USB, the code would be better factored if this logic was moved into `PS2KeyboardDevice` IMO.

Reply:
The problem is that there's a strong relation between scan codes to other attributes of the keyboard - like modifiers, which are currently tied quite strongly to given scan codes. I will try refactor it, but for actually getting somewhere in reasonable time, I think we should aim at getting USB keyboard support in and worry about abstractions later on.